crates/bevy_render/src/camera/visible_entities.rs

```diff
@@ -197,14 +214,89 @@ mod rendering_mask_tests {
     }
 }
 
+//Computes effective visibility for entities.
+//
+//In hirerarchies the actual visiblity of an entity isn't only the current value,
+//but also depends on ancestors of the entity./ To avoid traversing each hierarchy
+//and determine the effective visibility for each entity, this system listens to
+//visiblity and hierarchy changes and only then computes a value to be cached and
+//used by other systems.
+pub fn visible_effective_system(
+    children_query: Query<&Children>,
+    changes_query: Query<
+        (Entity, Option<&Parent>),
+        (With<Visible>, Or<(Changed<Visible>, Changed<Parent>)>),
+    >,
+    mut visible_query: Query<(&Visible, Option<&mut VisibleEffective>)>,
+    mut commands: Commands,
+) {
+    fn update_effective(
+        entity: Entity,
+        is_visible_parent: bool,
+        children_query: &Query<&Children>,
+        mut visible_query: &mut Query<(&Visible, Option<&mut VisibleEffective>)>,
+        mut commands: &mut Commands,
+    ) {
+        if let Ok((visible, maybe_visible_effective)) = visible_query.get_mut(entity) {
+            let is_visible = visible.is_visible & is_visible_parent;
+            if let Some(mut visible_effective) = maybe_visible_effective {
+                visible_effective.is_transparent = visible.is_transparent;
+
+                if visible_effective.is_visible == is_visible {
+                    return;
+                }
+
+                visible_effective.is_visible = is_visible;
+            } else {
+                commands.entity(entity).insert(VisibleEffective {
+                    is_visible,
+                    is_transparent: visible.is_transparent,
+                });
+            }
+
+            if let Ok(children) = children_query.get(entity) {
+                for child in children.iter() {
+                    update_effective(
+                        *child,
+                        is_visible,
+                        &children_query,
+                        &mut visible_query,
+                        &mut commands,
+                    );
+                }
+            }
+        }
+    }
+
+    for (entity, maybe_parent) in changes_query.iter() {
```

Comment:
This will end up being extremely expensive when constructing hierarchies (or when many nodes change on a given update). When all nodes have changed, it will walk each entity in the hierarchy's "subtree" (which is exponential). We need to do this in linear time for it to be workable (aka visit each node exactly once).

Reply:
It's not obvious, but the code already handles that. The following lines check that the current entity as no parent or has an effective visibility already. Therefore, for a newly created subtree only the topmost node satisfies the condition.

Reply:
Though on it a bit and what can be done here is a better filter for `changes_query`, so an entity will be included if:
* its visibility changed;
* its parent changed and it already has effective visibility.